sets/src/index.ts

```diff
@@ -0,0 +1,878 @@
+    if (!sets || (gen < 7 && CURRENT_ONLY.includes(format as Format))) continue;
+    for (const pokemon in sets) {
+      legacy[pokemon] = legacy[pokemon] || {};
+      for (const name in sets[pokemon]) {
+        legacy[pokemon][`${LEGACY[format]} ${name}`] = sets[pokemon][name];
+      }
+    }
+  }
+
+  return legacy;
+}
+
+async function importSets(
+  pokemon: string,
+  gen: calc.Generation,
+  setsByFormat: { [format: string]: PokemonSets },
+  numByFormat: { [format: string]: number }
+) {
+  const analysesByFormat = await getAnalysesByFormat(pokemon, gen);
+  if (!analysesByFormat) return;
+
+  for (const [f, analyses] of analysesByFormat.entries()) {
+    const { format } = Formats.fromString(f);
+    if (!format) continue;
+
+    let setsForPokemon = setsByFormat[format];
+    if (!setsForPokemon) {
+      setsForPokemon = {};
+      setsByFormat[format] = setsForPokemon;
+    }
+    let sets = setsForPokemon[pokemon];
+    if (!sets) {
+      sets = {};
+      setsForPokemon[pokemon] = sets;
+    }
+
+    for (const analysis of analyses) {
+      for (const moveset of analysis.movesets) {
+        const set = toPokemonSet(format, pokemon, moveset);
+        if (validPokemonSet(gen, format, set)) {
+          if (filter(format, pokemon, set)) continue;
+          sets[cleanName(moveset.name)] = set;
+          numByFormat[format] = (numByFormat[format] || 0) + 1;
+        } else {
+          invalid(gen, format, pokemon, moveset.name, set);
+        }
+      }
+    }
+  }
+}
+
+const SMOGON_ALIASES: { [id: string]: string } = {
+  darmanitanz: 'Darmanitan-Zen',
+  meloettap: 'Meloetta-Pirouette',
+  aegislashshield: 'Aegislash',
+  pumpkabooaverage: 'Pumpkaboo',
+};
+
+async function getAnalysesByFormat(pokemon: string, gen: calc.Generation) {
+  const id = toID(pokemon);
+  if (id.endsWith('totem')) return undefined;
+
+  pokemon = SMOGON_ALIASES[id] || pokemon;
+  const url = smogon.Analyses.url(pokemon, gen);
+
+  const retries = 3;
+  const wait = 40;
+  const retry = async (attempt = 0): Promise<Map<string, smogon.Analysis[]> | undefined> => {
+    try {
+      return smogon.Analyses.process(await request(url));
+    } catch (err) {
+      attempt++;
+      if (attempt > retries) throw err;
+      const timeout = Math.round(attempt * wait * (1 + Math.random() / 2));
+      warn(`Retrying ${url} in ${timeout}ms (${attempt}):`, err);
+      return new Promise(resolve => {
+        setTimeout(async () => {
+          resolve(await retry(attempt++));
+        }, timeout);
+      });
+    }
+  };
+
+  const analysesByFormat = retry();
+  if (!analysesByFormat) {
+    error(`Unable to process analysis: ${url}`);
+    return undefined;
+  }
+
+  return analysesByFormat;
+}
+
+function cleanName(name: string) {
+  return name.replace(/"/g, `'`);
+}
+
+function filter(format: Format, pokemon: string, set: PokemonSet) {
+  const hasMove = (m: string) => set.moves && set.moves.includes(m);
+  if (format === 'BH') return false;
```

Comment:
I can't find anything about that on Smogon, but maybe I'm misreading or looking in the wrong places. @TheImmortal for the official ruling :)

@Honko - does your application handle BH M-Ray/P-Don? What does your condition here look like?

Reply:
Mine still has a check for Gen6 BH to handle P-Don and P-Kyogre, but that condition never applies anymore since I stopped importing Gen6 BH sets once it was no longer the current gen. I never added a new check for Gen7 BH, probably because the bans happened after I had stopped really paying attention.

```
    private static boolean isValidSetForAlt(
            final Moveset moveset,
            final String altName,
            final Metagame metagame) {
        // everything is legal in BH except Primal Kyogre/Groudon without appropriate Orb in XY
        if (metagame.getTier() == BH) {
            return !(metagame.getGeneration() == XY
                    && altName.contains("Primal")
                    && !(moveset.getItems().contains("Red Orb") || moveset.getItems().contains("Blue Orb")));
        }
        // ...
        // all the other forme checks you already have
        // ...
    }
```

Reply:
> I can't find anything about that on Smogon, but maybe I'm misreading or looking in the wrong places. @TheImmortal for the official ruling :)

Groudon-Primal and Rayquaza-Mega in their raw formes are banned. Base Groudon with Red Orb and Rayquaza with Dragon Ascent are legal.

Reply:
Thanks @TheImmortal. Possibly worth clarifying on https://www.smogon.com/dex/sm/formats/bh/ & https://www.smogon.com/dex/xy/formats/bh/?